ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala

```diff
@@ -4,54 +4,204 @@
 package com.daml.platform.server.api.validation
 
 import com.daml.ledger.api.domain.LedgerId
+import com.daml.ledger.grpc.GrpcStatuses
 import com.daml.platform.server.api.ApiException
-import io.grpc.{Status, StatusRuntimeException}
-
+import com.google.protobuf.{Any => AnyProto}
+import com.google.rpc.{ErrorInfo, Status}
+import io.grpc.Status.Code
+import io.grpc.StatusRuntimeException
+import io.grpc.protobuf.StatusProto
 import scalaz.syntax.tag._
 
 trait ErrorFactories {
 
-  def ledgerIdMismatch(expected: LedgerId, received: LedgerId): StatusRuntimeException =
+  import ErrorFactories._
+
+  lazy val DuplicateCommandException: StatusRuntimeException =
     grpcError(
-      Status.NOT_FOUND.withDescription(
-        s"Ledger ID '${received.unwrap}' not found. Actual Ledger ID is '${expected.unwrap}'."
-      )
+      Status
+        .newBuilder()
+        .setCode(Code.ALREADY_EXISTS.value())
+        .setMessage("Duplicate command")
+        .addDetails(definiteAnswers(true))
+        .build()
     )
 
-  def missingField(fieldName: String): StatusRuntimeException =
-    grpcError(Status.INVALID_ARGUMENT.withDescription(s"Missing field: $fieldName"))
-
-  def invalidArgument(errorMsg: String): StatusRuntimeException =
-    grpcError(Status.INVALID_ARGUMENT.withDescription(s"Invalid argument: $errorMsg"))
+  def notFound(message: String): StatusRuntimeException =
+    grpcError(
+      Status
+        .newBuilder()
+        .setCode(Code.NOT_FOUND.value())
+        .setMessage(message)
+        .build()
+    )
 
-  def invalidField(fieldName: String, message: String): StatusRuntimeException =
-    grpcError(Status.INVALID_ARGUMENT.withDescription(s"Invalid field $fieldName: $message"))
+  /** @param expected Expected ledger id.
+    * @param received Received ledger id.
+    * @param definiteAnswer A flag that says whether it is a definite answer. Provided only in the context of command deduplication.
+    * @return An exception with the [[Code.NOT_FOUND]] status code.
+    */
+  def ledgerIdMismatch(
+      expected: LedgerId,
+      received: LedgerId,
+      definiteAnswer: Option[Boolean],
```

Comment:
Minor: I think we can safely assume that a wrong ledger ID can never be a definite answer. So IMO it would make sense to not even offer the possibility to return a `definite_answer = true` field here.

Reply:
Meta-comment: While it makes sense to require a `definite_answer` metadata field on all gRPC status codes other than `OK` for command submissions to facilitate the migration and testing the migration, in the long run, I'd argue that we could say that if there's no definite answer, then that's equivalent to `definite_answer = false`. But that's something for another discussion.